mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java

```diff
@@ -0,0 +1,168 @@
+        }
+    }
+
+    private Task.Result handleFullReindexingIterations(List<MailboxPath> mailboxPaths, GlobalRegistration globalRegistration,
+                                                       ReprocessingContext reprocessingContext) {
+        return mailboxPaths.stream()
+            .map(globalRegistration::getPathToIndex)
+            .flatMap(OptionalUtils::toStream)
+            .map(path -> {
+                try {
+                    return reIndex(path, reprocessingContext);
+                } catch (Throwable e) {
+                    LOGGER.error("Error while proceeding to full reindexing on {}", path, e);
+                    return Task.Result.PARTIAL;
+                }
+            })
+            .reduce(Task::combine)
+            .orElse(Task.Result.COMPLETED);
+    }
+
+    private Task.Result reIndex(MailboxPath path, MailboxSession mailboxSession, ReprocessingContext reprocessingContext) throws MailboxException {
+        MailboxRegistration mailboxRegistration = new MailboxRegistration(path);
+        LOGGER.info("Intend to reindex {}",path);
+        Mailbox mailbox = mailboxSessionMapperFactory.getMailboxMapper(mailboxSession).findMailboxByPath(path);
+        messageSearchIndex.deleteAll(mailboxSession, mailbox);
+        mailboxManager.addListener(path, mailboxRegistration, mailboxSession);
+        try {
+            return Iterators.toStream(
+                mailboxSessionMapperFactory.getMessageMapper(mailboxSession)
+                    .findInMailbox(mailbox, MessageRange.all(), MessageMapper.FetchType.Metadata, NO_LIMIT))
+                .map(MailboxMessage::getUid)
+                .map(uid -> handleMessageReIndexing(mailboxSession, mailboxRegistration, mailbox, uid))
+                .peek(reprocessingContext::updateAccordingToReprocessingResult)
+                .reduce(Task::combine)
+                .orElse(Task.Result.COMPLETED);
+        } finally {
+            LOGGER.info("Finish to reindex {}", path);
+            mailboxManager.removeListener(path, mailboxRegistration, mailboxSession);
+        }
+    }
+
+    private Task.Result handleMessageReIndexing(MailboxSession mailboxSession, MailboxRegistration mailboxRegistration, Mailbox mailbox, MessageUid uid) {
+        try {
+            Optional<ImpactingMessageEvent> impactingMessageEvent = findMostRelevant(mailboxRegistration.getImpactingEvents(uid));
+
+            Optional.of(uid)
+                .filter(x -> !wasDeleted(impactingMessageEvent))
+                .flatMap(Throwing.function(mUid -> fullyReadMessage(mailboxSession, mailbox, mUid)))
+                .map(message -> messageUpdateRegardingEvents(message, impactingMessageEvent))
+                .ifPresent(Throwing.consumer(message -> messageSearchIndex.add(mailboxSession, mailbox, message)));
+            return Task.Result.COMPLETED;
+        } catch (Exception e) {
+            LOGGER.warn("ReIndexing failed for {} {}", mailbox.generateAssociatedPath(), uid, e);
```

Comment:
Adding an exception in _slf4j_ API is working only if you have at most one other parameter.
`void warn(String msg, Throwable t)`
`void warn(String format, Object arg1, Object arg2)`

Reply:
no, I think it's working as expected (confirmed by @jeantil if I remember correctly), that's a hidden feature :)

Reply:
Never use hidden feature, as it may break silently.

Reply:
the alternative is highly verbose :/

Reply:
It's not hidden, just not type-checked